qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp

```diff
@@ -4216,20 +4216,29 @@ QByteArray VulkanPipelineStateViewer::ReconstructSpecializationData(const VKPipe
 {
   bytebuf specData;
 
-  // reconstruct the original spec data as best as we can
+  // We don't have access to the buffers in the original creation info, so we try to reconstruct
+  // from our preprocessed pipeline state instead. Note that this data might have a different order
+  // from the original call or have unused entries eliminated based on shader reflection.
   const bytebuf &src = sh.specializationData;
 
   for(size_t i = 0; i < mapEntries->NumChildren(); i++)
   {
     const SDObject *map = mapEntries->GetChild(i);
 
-    size_t srcByteOffset = map->FindChild("constantID")->AsUInt32() * sizeof(uint64_t);
     size_t dstByteOffset = map->FindChild("offset")->AsUInt32();
     size_t size = map->FindChild("size")->AsUInt32();
+    specData.resize_for_index(dstByteOffset + size - 1);
 
+    uint32_t constantId = map->FindChild("constantID")->AsUInt32();
+    int32_t idx = sh.specializationIds.indexOf(constantId);
+    if(idx == -1)
+    {
+      qWarning("Couldn't find original data for spec ID %u, substituting with zero", constantId);
+      continue;
+    }
+    size_t srcByteOffset = idx * sizeof(uint64_t);
```

Comment:
This packing shouldn't be assumed here, the offset in the shader's `specializationData` in the pipeline state is given by the `byteOffset` member of the shader constant. The array is parallel to the `specializationIds` you've added, so you can look up the corresponding constant directly in the reflection data. That way if the packing changes, the data is still found.